### ISSUE TITLE
IdnMapping.GetAscii is much faster in GlobalizationMode.Invariant

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#37350]
Is there a good reason to use the NLS/ICU implementations for IdnMapping rather than the managed implementation?  The managed implementation used for GlobalizationMode.Invariant appears to be much faster (and it appears there's more optimization opportunity, in particular around allocation, e.g. there's a StringBuilder that's always allocated and used).  Or if we need to use the native implementations for some reason, could we add the managed fast-path for avoiding the native implementation if no conversion is needed?

Invariant Off - ICU
|         Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| GetAscii_Lower | 240.1 ns | 0.86 ns | 0.72 ns |      - |     - |     - |         - |
| GetAscii_Upper | 241.8 ns | 1.30 ns | 1.01 ns |      - |     - |     - |         - |
| GetAscii_Emoji | 563.5 ns | 1.29 ns | 1.08 ns | 0.0076 |     - |     - |      48 B |

Invariant Off - NLS
|         Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| GetAscii_Lower | 314.0 ns | 1.05 ns | 0.93 ns |      - |     - |     - |         - |
| GetAscii_Upper | 318.6 ns | 1.04 ns | 0.97 ns |      - |     - |     - |         - |
| GetAscii_Emoji | 619.4 ns | 0.94 ns | 0.78 ns | 0.0076 |     - |     - |      48 B |

Invariant On
|         Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
| GetAscii_Lower |  20.64 ns | 0.036 ns | 0.028 ns |      - |     - |     - |         - |
| GetAscii_Upper |  20.03 ns | 0.031 ns | 0.028 ns |      - |     - |     - |         - |
| GetAscii_Emoji | 231.95 ns | 1.632 ns | 1.274 ns | 0.0408 |     - |     - |     256 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Globalization;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private readonly IdnMapping _idn = new IdnMapping();

    [Benchmark] public string GetAscii_Lower() => _idn.GetAscii("lowercase");
    [Benchmark] public string GetAscii_Upper() => _idn.GetAscii("UPPERCASE");
    [Benchmark] public string GetAscii_Emoji() => _idn.GetAscii("\xD83D\xDE00\xD83D\xDE00\xD83D\xDE00\xD83D\xDE00");
}
```

**Comment by safern** [comment]
Based on: https://github.com/dotnet/runtime/issues/37349#issuecomment-638332697-- I believe is for correctness, I would expect in Invariant mode we don't consider a lot of cases and we can't handle all the supported scenarios.

**Comment by tarekgh** [comment]
@stephentoub  to handle IDN correctly we have to have the normalization data (which is big and not trivial). Depending on ICU/NLS give us this functionality for free and even we don't have to worry about updating the data nor the IDN behavior moving forward. When enabling Invariant mode we don't promise the behavior will be correct or even updated to support more recent IDN features. This is not different than any other Globalization feature we lose when enabling Invariant mode. I am going to close this one as "by design".  please let me know if you think otherwise.

**Comment by tarekgh** [comment]
> "Or if we need to use the native implementations for some reason, could we add the managed fast-path for avoiding the native implementation if no conversion is needed?"

I think you mean in cases of Ascii strings. this could be possible but we need to revisit the managed implementation to know it still conform to the specs as this managed implementation is very old.

**Comment by stephentoub** [comment]
> I think you mean in cases of Ascii strings

Yes, which I'd expect to be a very common case.  Let me know if you disagree.  Just based on the numbers above, it seems like worst case we'd be adding 10% overhead when the string isn't ASCII, in exchange for saving 90% when it is.  Assuming those numbers are in the right ballpark and ASCII is as common as I'd expect, that feels like a good tradeoff.

**Comment by tarekgh** [comment]
@stephentoub I agree we should give it a try. I see the benefit for this optimization. Are you talking about GetAscii only or about GetUnicode. too? does the HTTP using both GetAscii and GetUnicode? I am asking because GetUnicode will need extra work to convert to Unicode first and then check the resulted string which may add some perf cost on non-Ascii cases but it is possible to do it too.

**Comment by stephentoub** [comment]
Thanks.  GetAscii is the more impactful API.

**Comment by tarekgh** [comment]
moved it to 6.0 but we can bring back to 5.0 if we get chance to look at it.
